### ISSUE TITLE
base_path option does not seem to work for unix domain sockets

### ISSUE THREAD

**Issue by hadmut** [HoneyryderChuck/httpx#64]
Hi, 

I'm looking for a way to do HTTP requests and use the faraday rest client gem with REST API servers on unix domain sockets (e.g. LXD, snapd, docker are offering REST APIs over unix domain sockets), and came to httpx, since httpx has support for unix domain sockets. 

But the  base_path option does not seem to work with unix domain sockets. 

Test 1:
require 'httpx'
socketpath = "/var/snap/lxd/common/lxd/unix.socket"
http = HTTPX.with(transport: "unix", addresses: [socketpath], base_path: "http://example")

response = http.get("http://xyz/")
works as expected, but it's a bit awkward. 

response = http.get("/")
results in 
httpx-1.3.1/lib/httpx/request.rb:88:in `initialize': invalid URI: / (HTTPX::Error)

shouldn't this – following the documentation – work by beeing appended to the base_path option value?

Test 2:
require 'faraday'
require "httpx/adapters/faraday"

client = Faraday.new do |builder|
  builder.adapter :httpx, transport: "unix", addresses: [socketpath]  , base_path: "http://example"
 end

again, 
p client.get("http://example/")
works, but 
p client.get("/")
doesn't, but this time with a different error, i.e.
httpx-1.3.1/lib/httpx/headers.rb:114:in `block in each': undefined method `empty?' for nil:NilClass (NoMethodError)

        yield(field, value) unless value.empty?
                                        ^^^^^^^

Debugging the values:
Extra_headers: {"connection"=>"keep-alive", "host"=>nil}

It fails when testin "host" for .empty?  
which isn't defined for nil, and thus a bug in any way. 

regards

**Comment by HoneyryderChuck** [comment]
Hi @hadmut ,

You want to use :origin, not :base_path. Can you confirm that works?

**Comment by hadmut** [comment]
Yes and no. 

The given example Test 1 (using httpx directly) works with using :origin, and even together with :base_path. 

But the given example Test 2 (using faraday with the httpx plugin) still results in 

httpx-1.3.1/lib/httpx/headers.rb:114:in `block in each': undefined method `empty?' for nil:NilClass (NoMethodError)


client = Faraday.new do |builder|
 builder.adapter :httpx, transport: "unix", addresses: [socketpath]  , origin: "http://dummy/"
 end
 p client.get("/")

**Comment by HoneyryderChuck** [comment]
Would you mind running your test 2 example with HTTPX_DEBUG=2 enviar, then posting the resulting debug logs here? Alternatively, you could also run it against a local httpx cloned repo and add a binding.pry statement in the line where the exception is raised, to figure out which header key is being assigned to a a nil value.

**Comment by HoneyryderChuck** [comment]
Nevermind,  you posted earlier. I think I know what's going on: [when building requests in faraday](https://github.com/HoneyryderChuck/httpx/blob/599b6865da1df7331dd52b7963e60cdbe3370128/lib/httpx/adapters/faraday.rb#L67), the session options aren't being passed downstream. I'll patch it in the next few days.

**Comment by HoneyryderChuck** [comment]
Closing as not a bug.

The `:origin` and `:base_path` do not work as expected under Faraday, as they collide with Faraday specific features. Namely, in your example, you should create your Faraday client with a default base uri, if you want to use relative URIs. It's something that Faraday documents somewhere for sure.

```ruby
- client = Faraday.new do |builder|
+ client = Faraday.new("http://dummy/") do |builder|
-  builder.adapter :httpx, transport: "unix", addresses: [socketpath] , origin: "http://dummy/"
+  builder.adapter :httpx, transport: "unix", addresses: [socketpath]
end
p client.get("/")
```

The error doesn't help due to a combination of factors: when not passed a default base URL to merge with relative URLs, Faraday will default to using `http` scheme internally, and it'll pass, as per you example above, the `http:/get` URI downstream; this URL does not contain a host, and `httpx` should have flagged it as an invalid URI, however it was using `URI::HTTP#relative?`, which checks if the URI has a scheme, but not if it has a host. This has been fixed in 02ed2ae87dde1ca98bd8d76a67a4113d30f20de9 .
